### ISSUE TITLE
uprobe with custom address

### ISSUE THREAD

**Issue by atwufei** [iovisor/bcc#1953]
trace.py only traces functions' entry/exit, will it add support for tracing custom addresses in the function? And it's possible to read/write the local variables?

**Comment by yonghong-song** [comment]
The python API `attach_kprobe` does support `symbol+offset` and `attach_uprobe` does support arbitrary address. It is possible to trace `func+offset` in trace.py, probably you can extend the interface to allow `func_name[(...)]+offset`?

Regarding to read/write local variables, if local variable is in register, you cannot. if local variable is in stack, you may use `bpf_probe_write_user` helper. You need to dig out the dwarf info itself to figure out how to access a particular local variable at which insn offset.

**Comment by atwufei** [comment]
Thank you Yonghong. I made a dirty patch to trace.py, and it works with the custom address.

If local variable is in register, it's not even readable?

If local variable is in stack, can I access it by symbol or its address? Is there any example I can follow?

**Comment by palmtenor** [comment]
I think for local variables in registers, @yonghong-song is referring to write. You can always read a register value, you just need to be sure that the value you care about is indeed in that register.

If local variable is in stack, you can access it if you are sure about its offset using `bpf_probe_read`. BPF does not have the concept of symbol so you can not access using it.

It seems like what you need is exactly what USDT provides: probe into a specific location that is neither function entry nor exit, and you want to ensure the values you cared about are put into known register or memory location~

**Comment by atwufei** [comment]
Thank you @palmtenor. USDT is a good solution but defined statically, it might be not the case during debugging. In Systemtap, it's able to use symbol directly, I am not sure if it's possible to add this feature into bcc.

**Comment by palmtenor** [comment]
I'm not sure how SystemTap does it, but theoretically if we can parse DWARF data to know the exact stack and register layout at particular instruction, it could be used in BPF program to read those values. I think is unlikely that BPF Kernel side would ever support DWARF parsing, so this would need to be done in user space and pass the layout like register name / memory offset to BPF program, similar to how USDT works.

**Comment by atwufei** [comment]
Agreed. I don't think it's a good idea to parse DWARF in kernel side either. But if the userspace side supports it, it will be much more convenient. e.g. systemtap can use the c++ namespace in the symbol, such as "ClassA::func", but afaik, I should get the demangled symbol first in order to use trace.py or similar.

**Comment by kroemeke** [comment]
regarding uprobe DWARF - I have a PoC using pyelftools linked somewhere here https://github.com/iovisor/bcc/issues/1803 , but doesn't support local variables - what I needed at work was specifically traversing some fat structs so it is just calculating struct offsets.

**Comment by fche** [comment]
In case it helps, systemtap's symbol-resolution logic allows it to compute source-code based statement-granularity probe addresses, and access local/global variables, much as a full symbolic debugger does.  This requires a considerable mass of code that parses DWARF, and is unfortunately not packaged in a form that another project like bcc could reuse.  See the `loc2stap.cxx` and the monster `tapset.cxx` modules.
